pkg/job/job.go

```diff
@@ -14,21 +16,58 @@ import (
 	"github.com/insomniacslk/xjson"
 )
 
-// JobDescriptor models the JSON encoded blob which is given as input to the
-// job creation request. A JobDescriptor embeds a list of TestDescriptor.
-type JobDescriptor struct {
+// Descriptor models the deserialized version of the JSON text given as
+// input to the job creation request.
+type Descriptor struct {
 	JobName         string
 	Tags            []string
 	Runs            uint
 	RunInterval     xjson.Duration
-	TestDescriptors []*test.TestDescriptor
+	TestDescriptors []test.Descriptor
 	Reporting       Reporting
 }
 
+// Validate performs sanity checks on the job descriptor
+func (d *Descriptor) Validate() error {
+
+	if len(d.TestDescriptors) == 0 {
+		return errors.New("need at least one TestDescriptor in the JobDescriptor")
+	}
+	if d.JobName == "" {
+		return errors.New("job name cannot be empty")
+	}
+	if d.RunInterval < 0 {
+		return errors.New("run interval must be non-negative")
+	}
+
+	if len(d.Reporting.RunReporters) == 0 && len(d.Reporting.FinalReporters) == 0 {
+		return errors.New("at least one run reporter or one final reporter must be specified in a job")
+	}
+	for _, reporter := range d.Reporting.RunReporters {
+		if strings.TrimSpace(reporter.Name) == "" {
+			return errors.New("run reporters cannot have empty or all-whitespace names")
+		}
+	}
+	return nil
+}
+
+// ExtendedDescriptor is a job descriptor which has been extended with the full
+// description of the test and cleanup steps obtained from the test fetcher.
+type ExtendedDescriptor struct {
+	Descriptor
+	StepsDescriptors []test.StepsDescriptors
```

Comment:
Why is this an array of multiple`StepsDescriptors`? Doesn't one `StepsDescriptors` already contain multiple Test and Cleanup Steps?

Reply:
I read this PR long time ago and can remember things wrong, but IIRC one `test.StepsDescriptors` is a set of steps for only one test (in terms of JobDescription), while `ExtendedDescriptor` contains all tests.

Reply:
Yeah, what @xaionaro said, this `[]test.StepsDescriptors` is rebuilt from the serialization in the db which contains, for every `Test`, all the `StepsDescriptors`.